File: scrape.py
from time import sleep
import httpx
import json
import sys
from bs4 import BeautifulSoup
from urllib.parse import urlparse, parse_qs

proxies = [
    None,
    '190.26.201.194:8080',
    '164.70.72.55:3128',
    '52.183.8.192:3128',
    '162.55.84.170:3128',
    '201.184.171.244:999'
]
base_url = "https://asianembed.io"


def fetchContent(url):
    while True:
        for proxy in proxies:
            try:
                return httpx.get(url=url, follow_redirects=False, proxies=proxy).text
            except Exception:
                print('Switching proxy server %s.' % proxy)
                sleep(5)
                continue


def fetchMovies(page):
    content = fetchContent(base_url + '/?page=%s' % page)

    soup = BeautifulSoup(content, 'lxml')

    # list of movies
    movies = soup.find_all('li', class_='video-block')

    print('Found %s movies. #%s' % (len(movies), page))

    if (len(movies) <= 0):
        raise Exception('No movies found for this page #%s.' % page)

    for movie in movies:
        movieLink = base_url + movie.find('a')['href']
        content = fetchContent(movieLink)
        soup = BeautifulSoup(content, 'lxml')
        main = soup.find('div', class_='video-info-left')

        q = parse_qs(
            urlparse(main.find('div', class_='play-video').find('iframe')['src']).query)
        title = q['title'][0]
        cover = q['cover'][0]

        print('Start scraping %s movie.' % title)
        movieDetail = {
            'title': title,
            'cover': cover,
            'description': '',
            'episodes': []
        }

        episodes = main.find_all('li', class_='video-block')

        print('Found %s episodes from %s movie.' %
              (len(episodes), q['title'][0]))

        for episode in episodes:
            episodeLink = base_url + episode.find('a')['href']
            content = fetchContent(episodeLink)
            soup = BeautifulSoup(content, 'lxml')
            main = soup.find('div', class_='video-info-left')
            embed = main.find('div', class_='play-video').find('iframe')
            description = main.find(
                'div', class_='video-details').find('div', class_='post-entry').text.strip()

            q = parse_qs(urlparse(embed['src']).query)

            image = episode.find('div', class_='img').find('img')
            date = episode.find('span', class_='date').text
            slug = episode.find('a')['href'].split('/')[-1]
            name = episode.find('div', class_='name').text.strip()
            print('Start scraping %s.' % name)
            episodeDetail = {
                'id': q['id'][0],
                'embed': embed['src'],
                'typesub': q['typesub'][0],
                'slug': slug,
                'number': slug.split('-')[-1],
                'image': {
                    'alt': image['alt'],
                    'src': image['src']
                },
                'name': name,
                'meta': episode.find('div', class_='meta').text.strip(),
                'created_at': date,
                'updated_at': date
            }

            movieDetail['title'] = q['title'][0]
            movieDetail['cover'] = q['cover'][0]
            movieDetail['description'] = description
            movieDetail['episodes'].append(episodeDetail)

        with open('movies/%s-%s.json' % (title, cover), 'w') as f:
            f.write(json.dumps(movieDetail))
            f.close()
        print('Completed scraping %s movie.' % title)
    else:
        print('Completed scraping page #%s.' % page)


page = int(sys.argv[1]) or 1
endPage = int(sys.argv[2]) or 10

while page <= endPage:
    try:
        fetchMovies(page)
    except Exception:
        break
    page += 1


print("Scraping completed.")

File: scrape.php
<?php

require __DIR__ . '/vendor/autoload.php';

use App\Core\Application;
use Symfony\Component\DomCrawler\Crawler;

$urls = [
    '-s' => ['url' => 'https://asianembed.io', 'tag' => 'sub'],
    '-r' => ['url' => 'https://asianembed.io/recently-added-raw', 'tag' => 'raw'],
    '-m' => ['url' => 'https://asianembed.io/movies', 'tag' => 'movie'],
    '-k' => ['url' => 'https://asianembed.io/kshow', 'tag' => 'show'],
    '-p' => ['url' => 'https://asianembed.io/popular', 'tag' => 'popular'],
    '-o' => ['url' => 'https://asianembed.io/ongoing-series', 'tag' => 'ongoing']
];

class Scraper
{
    private $baseUrl = 'https://asianembed.io';
    // private $postUrl = 'https://ams.htetwunsan.com';
    private $postUrl = 'http://localhost:8001';

    private function getCrawler(string $url): Crawler
    {
        while (true) {
            $content = file_get_contents($url);
            if ($content === false) {
                echo "Cannot fetch content from $url" . PHP_EOL;
                sleep(5);
                continue;
            }
            return new Crawler($content);
        }
    }

    private function getEpisodeSlugs(Crawler $crawler): array
    {
        return $crawler->filter('li.video-block')->each(function (Crawler $episode) {

            $slug = $episode->filter('a')->attr('href');

            return $slug;
        });
    }

    private function getEpisodeDetail(Crawler $crawler): array
    {
        $embed = $crawler->filter('div.play-video')->filter('iframe');
        $embed = $crawler->filter('div.play-video')->filter('iframe');
        $url = parse_url($embed->attr('src'));
        parse_str($url['query'], $q);
        $description = trim($crawler->filter('div.video-details')->filter('div.post-entry')->text());

        return [
            'id' => $q['id'],
            'embed' => $embed->attr('src'),
            'video' => [
                'title' => $q['title'],
                'cover' => $q['cover'],
                'description' => $description,
            ]
        ];
    }

    private function getAllEpisodes(Crawler $crawler): array
    {
        return $crawler->filter('li.video-block')->each(function (Crawler $episode) {
            $data = $this->getEpisodeNormalData($episode);
            $sub = trim($episode->filter('div.type')?->text()) == 'SUB';
            $data['sub'] = $sub;
            return $data;
        });
    }

    private function getEpisodeNormalData(Crawler $episode): array
    {
        $slug = $episode->filter('a')->attr('href');
        $explodedSlug = explode('-', $slug);
        $number = end($explodedSlug);
        $name = trim($episode->filter('div.name')->text());
        $image = $episode->filter('div.img')->filter('img');
        $date = trim($episode->filter('span.date')->text());
        $meta = trim($episode->filter('div.meta')->text());

        return [
            'slug' => $slug,
            'name' => $name,
            'number' => $number,
            'image' => [
                'src' => $image->attr('src'),
                'alt' => $image->attr('alt')
            ],
            'meta' => $meta,
            'date' => $date
        ];
    }

    private function getExistingEpisodes(string $videoCover, string $tag): array
    {
        $postData = http_build_query([
            'video_cover' => $videoCover,
            'tag' => $tag
        ]);
        $context = stream_context_create([
            'http' => [
                'method' => 'POST',
                'header' => 'Content-type: application/x-www-form-urlencoded',
                'content' => $postData
            ]
        ]);

        while (true) {
            $url = $this->postUrl . '/api/episodes/existing';
            $result = file_get_contents($url, false, $context);
            if ($result === false) {
                echo "Cannot post content to $url" . PHP_EOL;
                sleep(5);
            }
            break;
        }
        return json_decode($result, true);
    }

    private function uploadEpisode(array $data)
    {
        $postData = http_build_query($data);
        $context = stream_context_create([
            'http' => [
                'method' => 'POST',
                'header' => 'Content-type: application/x-www-form-urlencoded',
                'content' => $postData
            ]
        ]);
        while (true) {
            $url = $this->postUrl . '/api/episodes';
            $result = file_get_contents($url, false, $context);
            if ($result === false) {
                echo "Cannot post content to $url" . PHP_EOL;
                sleep(5);
            }
            break;
        }
        $result = json_decode($result, true);
        $resultEpisode = $result['episode'];
        echo "Upserting episode " . $data['name'] . " completed. Result: $resultEpisode" . PHP_EOL;
    }

    public function run(string $tag, string $url)
    {
        for ($page = 1; $page <= 10; ++$page) {
            $newUrl = $url . "?page=$page";

            echo "Start scraping $newUrl with tag $tag." . PHP_EOL;
            // crawling list
            $crawler = $this->getCrawler($newUrl);

            $slugs = $this->getEpisodeSlugs($crawler); // list of slugs from list page

            foreach ($slugs as $slug) {
                // crawling detail
                $crawler = $this->getCrawler($this->baseUrl . $slug)->filter('div.video-info-left');

                $allEpisodes = $this->getAllEpisodes($crawler);
                $allEpisodeCount = count($allEpisodes);

                $videoCover = $this->getEpisodeDetail($crawler)['video']['cover'];

                $existingEpisodeSlugs = array_flip($this->getExistingEpisodes($videoCover, $tag));

                foreach ($allEpisodes as $key => $episode) {
                    if (array_key_exists($episode['slug'], $existingEpisodeSlugs)) {
                        echo "Episode " . $episode['name'] . " is already existed. Skipping..." . PHP_EOL;
                        continue;
                    }
                    // crawling detail
                    $crawler = $this->getCrawler($this->baseUrl . $episode['slug'])->filter('div.video-info-left');

                    $episode = $this->getEpisodeDetail($crawler);
                    $episode['video']['episode_count'] = $allEpisodeCount;

                    $this->uploadEpisode(array_merge($episode, $allEpisodes[$key], ['tag' => $tag]));
                }
            }
        }
    }
}

$app = new Application(__DIR__);

(require base_path() . '/app/bootstrap.php')($app);

/**
 * @var Scraper $scraper
 */
$scraper = $app->get(Scraper::class);

$arg = $urls[$argv[1]];
$url = $arg['url'];
$tag = $arg['tag'];

$scraper->run($tag, $url);
